### ISSUE TITLE
Apply a build diff using prebuilt artifacts

### ISSUE THREAD

**Issue by jeremydw** [grow/grow#928]
Grow's `deploy` command currently works by building a fileset/manifest, and then "applying" the generated fileset against an existing fileset/manifest to know exactly which files to add, edit, and delete.

Currently, the "build" and "apply" steps are coupled. With build sharding (#908) we want to be able to produce a build independently (for example, in Google Cloud Build), upload the fileset/manifest corresponding to the git hash to a storage bucket (Google Cloud Storage), and then later be able to "apply" the fileset in a different environment without needing to rebuild the fileset.

Here's a sample workflow:

1. Developer makes a Git commit and pushes to GitHub.
1. Triggers a Google Cloud Build (and appropriate build sharding per #908).
1. GCB uploads the fileset for the commit hash to a predictable location in GCS (this can be a new setting in Grow's `podspec` akin to `deployments` – where the user can specify the bucket to upload to).

Now we have the entire fileset for the commit hash prebuilt.

Then, If we want to run the same build again, on another machine (i.e. for deployment):

1. Our new command to "apply" the fileset/manifest can be used.
1. In this scenario, instead of building the fileset all over again, the fileset will be downloaded from GCS.
1. Then, the manifests (from GCS and from the local destination) will be locally "applied" without having to fully rebuild everything.
1. The end result is the same as running `grow deploy` (but without rebuilding the whole site, since we've done that already in GCB).

Since this pattern could work with any Grow deployment, I think the ability to skip the local build and download the fileset/manifest (and then apply them against any deployment destination that might be already configured), should be a `grow deploy` flag.

Relevant code locations:

1. Once we download the build/manifest from GCS, we use the prebuilt GCS index and compare it against the index we're deploying to by creating a build __diff__ – [grow.deployments.Diff.create](https://github.com/grow/grow/blob/master/grow/deployments/indexes.py#L128).
1. We "apply" the diff by passing the just-created __diff__ and a map of paths to rendered files to [grow.deployments.Diff.apply](https://github.com/grow/grow/blob/master/grow/deployments/indexes.py#L205).

Our new flag could be used to alter the behavior of [grow.deployments.destinations.base.BaseDestination.deploy](https://github.com/grow/grow/blob/master/grow/deployments/destinations/base.py#L280) – to pull content from GCS (or a local temp folder that we download to, etc.) rather than rebuilding the whole site again.

**Comment by Zoramite** [comment]
Thanks for writing this all out. With the latest versions of the rendering pipeline it generates the `RenderedDocument`s for all the files that we render, which includes a hash of the file and a path to  an optional temporary file (so everything doesn't need to be kept in memory for larger builds).

I picture this apply feature being very similar to what the render pipeline does, like so:

- Make the routes tree of all of the paths like a normal build.
- Use the serving paths to determine what the final filenames are.
- Read the information on the file from whatever source (local, gcs, etc) to create a `RenderedDocument` including the hash of the file for use in the index.
- Proceed with normal deploy (Process already [uses `RenderedDocument`](https://github.com/grow/grow/blob/master/grow/deployments/destinations/google_cloud_storage.py#L112)).

So basically we should be able to make a separate render pipeline that does the lookup of the file information instead of doing the normal render and it should work.

Specifically, if we use create a subclass of the [`RenderDocumentController`](https://github.com/grow/grow/blob/master/grow/rendering/render_controller.py#L134) we could change the behaviour of the `render()` method to do a file lookup instead of the current jinja2 rendering.

I'll think it over some more, but it may just be as simple as overriding that one class and updating the logic to create that class instead of the normal classes when doing the rendering pipeline.
